server/client.go

```diff
@@ -1005,7 +1005,8 @@ func (c *client) flushClients(budget time.Duration) time.Time {
 			continue
 		}
 
-		if budget > 0 && cp.out.lft < 2*budget && cp.flushOutbound() {
+		// Try to flush in place, if producer and consumer are client.
+		if c.kind == CLIENT && cp.kind == CLIENT && budget > 0 && cp.out.lft < 2*budget && cp.flushOutbound() {
```

Comment:
No, we can't do that here in this function since it is called in various place with a budget where the expectation is that message is possibly flushed in place (or effort to do so is done). And it can be situations where the kind is not CLIENT (think internal send queues, JetStream, etc...).
To minimize risks, I would then limit the single call that we make in readLoop() where we pass a budget, and only that one behaving this way. This probably means having a function flushClients() that behaves the way it does today, and a different version that can have this check.
flushClients(budget) invokes flushClientsWithCheck(budget, clientsKindOnly) with clientsKindOnly==false.
Then the check could be:
```
if budget > 0 && (!clientsKindOnly || c.kind == CLIENT && cp.kind == CLIENT) && cp.out.lft...
```
Note that I am bad at naming things, so if you can think of a better function name and field, go for it :-)